pyogrio/_io.pyx

```diff
@@ -757,15 +796,24 @@ def ogr_read(
         layer = 0
 
     if fids is not None:
-        if where is not None or bbox is not None or skip_features or max_features:
+        if where is not None or bbox is not None or sql is not None or skip_features or max_features:
             raise ValueError(
-                "cannot set both 'fids' and any of 'where', 'bbox', "
+                "cannot set both 'fids' and any of 'where', 'bbox', 'sql', "
                 "'skip_features' or 'max_features'"
             )
         fids = np.asarray(fids, dtype=np.intc)
 
+    if sql is not None:
+        if where is not None or bbox is not None or fids is not None: # or skip_features or max_features:
+            raise ValueError(
+                "cannot set both 'sql' and any of 'where', 'bbox', 'fids'"
+            )
+
     ogr_dataset = ogr_open(path_c, 0, kwargs)
-    ogr_layer = get_ogr_layer(ogr_dataset, layer)
+    if sql is None:
```

Comment:
Defaulting the layer to 0 (lines 794-796) can move to this block since it is not used if `sql` is provided.